clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -1293,17 +1287,16 @@ public void setInflightCommitStatus(final boolean inflightCommitStatus) {
     }
 
     static class MemberInfo {
-        Optional<String> memberId;
+        String memberId = "";
```

Comment:
taking this a step further, I wonder if it would be clearer if this commitManager (and this MemberInfo), just take the memberId as constructor parameter. The memberId is generated by the membershipMgr once at startup, won't change, so it could pass it on to the commitManager upon creation here:
https://github.com/apache/kafka/blob/553e6b4c6d912b3b619b5a1a5263b616b6d3f62e/clients/src/main/java/org/apache/kafka/clients/consumer/internals/RequestManagers.java#L197 (creating membershipMgr first)
Then the `onMemberEpochUpdated` would only receive the epoch, which is what changes. What do you think?

Reply:
Yes, I think it's very ideal! But after quickly going through `RequestManagers`, it seems that `MembershipManager` relies on `CommitRequestManager`. If that's the case, it could be tricky.
https://github.com/apache/kafka/blob/553e6b4c6d912b3b619b5a1a5263b616b6d3f62e/clients/src/main/java/org/apache/kafka/clients/consumer/internals/RequestManagers.java#L197-L219

Reply:
uhm true. Then what about we simply do a kind of commitManager.setMemberId(memberId) from the membershipMgr constructor (right after assigning the commitRequestManager)? The point I'm after in the commitMgr is to clearly separate the `memberId` (set once, never updated) from the `memberEpoch` (may be updated and the commitMgr needs to know about those changes)

Reply:
I reviewed the code and found that in `AsyncKafkaConsumer`, the `updateGroupMetadata` method uses `memberId` to update `ConsumerGroupMetadata`[0]. Unfortunately, this method is passed as a lambda to implement `onMemberEpochUpdated`[1]. As a result, we can't remove the `memberId` parameter from `onMemberEpochUpdated` since `memberId` cannot be accessed directly within `AsyncKafkaConsumer`.

The best we could do now is to make sure we only set `MemberInfo.memberId` once at `membershipManager` constructor.

[0]
https://github.com/apache/kafka/blob/140d35c5459f4c7a91b23a238467cb5aa01b59fb/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L635-L646

[1]
https://github.com/apache/kafka/blob/140d35c5459f4c7a91b23a238467cb5aa01b59fb/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L354

It's pass as lambda to L159:
https://github.com/apache/kafka/blob/140d35c5459f4c7a91b23a238467cb5aa01b59fb/clients/src/main/java/org/apache/kafka/clients/consumer/internals/RequestManagers.java#L145-L159